gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/FlowCompilationValidationHelper.java

```diff
@@ -73,12 +74,12 @@ public class FlowCompilationValidationHelper {
 
   @Inject
   public FlowCompilationValidationHelper(Config config, SharedFlowMetricsSingleton sharedFlowMetricsSingleton,
-      UserQuotaManager userQuotaManager, FlowStatusGenerator flowStatusGenerator) {
```

Comment:
how do we know the `topologySpecFactory` has been initialized by now?

Reply:
TopologySpecFactory is guice based initialized

Reply:
Yes I see that but what is the `waiting` you refer to in description "making spec compiler as a topology spec catalog listener and then wait for topology spec catalog to populate topologies." How do you know the `topologySpecCatalog` is populated?

Reply:
Checking `GobblinServiceManager` start to see how `TopologySpecCatalog` is populated and it may not have the topologies loaded if it uses the listener method rather than `TopologySpecFactory.getTopologies()` https://github.com/apache/gobblin/blob/33660f5d03d703456be73e30ea80d3310457ac7d/gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java#L508

One safe guard we can add is both `Orchestrator` and `SpecCompiler`'s`onAddSpec` method should check `if specCompiler.isActive` is true before accepting specs to pass to compiler. https://github.com/apache/gobblin/blob/33660f5d03d703456be73e30ea80d3310457ac7d/gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java#L522

Reply:
SpecCompiler gets information about topology specs through orchestrator's onAddSpec. And orchestrator gets info about topology specs because it is a listener on topology spec catalog.
So compiler has to "wait for topology spec catalog to populate topologies". topology spec catalog is populated here https://github.com/apache/gobblin/blob/master/gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GobblinServiceManager.java#L508 but we may want compiler to have topology information before this code reaches.